### PR TITLE
Feature multi pointer input refactor

### PR DESCRIPTION
**Pull request by StephenHodgson**
Overview
---
Probably needs a bit more cleaning up but the basic functionality is there.

- [ ] Get Parabola and Line pointers to look/feel better
- [ ] New Pointer Teleport Manger doesn't work
- [ ] Xbox controller input refactor
- [ ] Spatial Mouse
- [x] Hardware Keyboard support
- [ ] There are three cursors in the scene. The Gaze cursor should disappear when a motion controller(s) is detected. (Gaze cursor should be displayed only when there are no motion controllers in the scene)

mrtk_multipointer1

- [ ] On thumbstick forward, the parabolic beam should be displayed without straight beam. (currently displayed together).

mrtk_multipointer2

- [ ] On thumbstick left/right/backward, the parabolic beam should not be displayed.
- [ ] Sometimes, only single controller visualizes the beam and cursor.
- [ ] Sometimes, a parabolic beam is displayed on trigger button event.
- [ ] Some NullReferenceExceptions

20180313_143326_mixedreality_moment 3

### CODE REVIEW

File: Assets/MixedRealityToolkit-Examples/Input/Scenes/InputManagerTest.unity [hunk 1]
```diff
@@ -113,86 +113,174 @@ NavMeshSettings:
     debug:
       m_Flags: 0
   m_NavMeshData: {fileID: 0}
---- !u!1 &107556006
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-Examples/Input/Scenes/InputManagerTest.unity]: I need to revert the changes to this scene.

File: Assets/MixedRealityToolkit/Utilities/Scripts/Attributes/RequireSingletonAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+#if UNITY_EDITOR
+using UnityEditor;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/Utilities/Scripts/Attributes/RequireSingletonAttribute.cs]: Looks like this class need to move into a Editor folder

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/MotionController/AttachToController.cs.meta [hunk 3]
```diff
@@ -1,6 +1,6 @@
 fileFormatVersion: 2
-guid: 4b8dbecb5fb4e0a4598eba9c83cb6b3c
-timeCreated: 1517585792
+guid: 9a5088063237d3944a918ae215b1f324
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/MotionController/AttachToController.cs.meta]: revert

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/MotionController/MotionControllerVisualizer.cs.meta [hunk 4]
```diff
@@ -1,6 +1,6 @@
 fileFormatVersion: 2
-guid: 9746d77301599614e874ca00077b49b3
-timeCreated: 1504899848
+guid: 03314f8008ae1ed4fbd45c95a7ed5182
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/MotionController/MotionControllerVisualizer.cs.meta]: revert

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Managers/MixedRealityTeleport.cs [hunk 5]
```diff
@@ -1,50 +1,41 @@
...
-#endif
 
 namespace MixedRealityToolkit.InputModule
 {
     /// <summary>
     /// Script teleports the user to the location being gazed at when Y was pressed on a Gamepad.
     /// </summary>
     [RequireComponent(typeof(SetGlobalListener))]
-    public class MixedRealityTeleport : Singleton<MixedRealityTeleport>, IControllerInputHandler
+    public class MixedRealityTeleport : Singleton<MixedRealityTeleport>, IInputHandler
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Managers/MixedRealityTeleport.cs]: We should mark this obsolete and point to the new teleport manager.

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Managers/InputManager.cs [hunk 6]
```diff
@@ -0,0 +1,1350 @@
...
+using UnityEngine.Windows.Speech;
+#endif
+
+namespace MixedRealityToolkit.InputModule
+{
+    /// <summary>
+    /// Input Manager is responsible for managing input sources and dispatching relevant events
+    /// to the appropriate input handlers. 
+    /// </summary>
+    public class InputManager : Singleton<InputManager>
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Managers/InputManager.cs]: move back into root Input Module folder.

**Comment by cre8ivepark** [comment]
Thanks, @StephenHodgson for sharing this!
Some updates needed for interaction behavior:

- New TeleportCursor.prefab appears on top of thumbstick. It should be displayed at the end of the parabolic beam.
- Old TeleportMarker.prefab should be removed. It appears at the end of the straight pointer beam.
- On thumbstick forward, the parabolic beam should be displayed without straight beam. (currently displayed together).
- On thumbstick left/right/backward, the parabolic beam should not be displayed.
- Randomly only single controller visualizes the beam and cursor.
- Some NullReferenceExceptions 

NullReferenceException: Object reference not set to an instance of an object
MixedRealityToolkit.InputModule.Pointers.TeleportPointer.OnInputReleased () (at Assets/MixedRealityToolkit/InputModule/Scripts/Pointers/TeleportPointer.cs:171)
MixedRealityToolkit.InputModule.Pointers.ParabolicTeleportPointer.OnInputReleased () (at Assets/MixedRealityToolkit/InputModule/Scripts/Pointers/ParabolicTeleportPointer.cs:62)
MixedRealityToolkit.InputModule.Pointers.TeleportPointer.Update () (at Assets/MixedRealityToolkit/InputModule/Scripts/Pointers/TeleportPointer.cs:79)

NullReferenceException: Object reference not set to an instance of an object
MixedRealityToolkit.InputModule.Pointers.ParabolicTeleportPointer.OnSelectPressed () (at Assets/MixedRealityToolkit/InputModule/Scripts/Pointers/ParabolicTeleportPointer.cs:57)
MixedRealityToolkit.InputModule.Pointers.TeleportPointer.Update () (at Assets/MixedRealityToolkit/InputModule/Scripts/Pointers/TeleportPointer.cs:71)

multipointer_issue1

**Comment by StephenHodgson** [comment]
Looks like something got broken with the GUIDs for the teleport cursors. I'll be sure to check on that later today.

**Comment by StephenHodgson** [comment]
@cre8ivepark you're more than welcome to open PRs against this branch. It lives in the [main MRTK repo](https://github.com/Microsoft/MixedRealityToolkit-Unity/tree/Feature-MultiPointerInputRefactor)

**Comment by Railboy** [comment]
@StephenHodgson I'll handle cleaning this up from here. @cre8ivepark let me know if you have any style requests.

**Comment by StephenHodgson** [comment]
Yeah we need to remove the old teleport logic and prefab

**Comment by Railboy** [comment]
The MRTK editors were previously ensuring that these null references wouldn't happen so I've created some custom inspectors to restore them. However, I've put them in MixedRealityToolkit/InputModule/Scripts/Editor/LinePointerEditors.cs so we can side-step the whole issue of keeping these outside of Editor folders.

@StephenHodgson which scene are you using to stage pointers? I'm looking in MixedRealityToolkit-Examples/Input/Scenes/MotionControllerTest and I don't see any pointer prefabs in there.

**Comment by StephenHodgson** [comment]
Checkout the InputManager. On the InteractionInputSources component there's fields add pointer prefabs. They should auto populate in any scene with the default InputManager prefab.

**Comment by StephenHodgson** [comment]
@cre8ivepark So the null ref errors you were getting was because the new teleport manager wasn't in the scene. I'll be sure to update the Camera Prefab to have it.

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/MotionController/ControllerFinder.cs [hunk 7]
```diff
@@ -1,6 +1,7 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
+using MixedRealityToolkit.Utilities.Attributes;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/MotionController/ControllerFinder.cs]: @DDReaper still not sure I like how we've split scripts into Utilities and Common.

**Comment by Railboy** [comment]
>Checkout the InputManager. On the InteractionInputSources component there are fields to add pointer prefabs. They should auto populate in any scene with the default InputManager prefab and motion controllers detected.

Ah, interesting. I like this setup, but it makes it difficult to do a couple of important things:

- Coordinate the behavior of multiple interacting pointers (eg line pointer selection + parabolic pointer teleportation). For instance, pointers have an array called DisableWhenActive that can be populated with other pointers. This can prevent, say, the line pointer from being active while the teleportation pointer is active. But this relationship has to be set up explicitly. (This is why @cre8ivepark was seeing overlapping pointers.)

- Specifying controller element and handedness manually. (This is why @cre8ivepark was seeing the pointer appear on the thumbstick instead of on the pointer element - it looks like it wasn't configured.)

 - Turn sets of pointers on / off when switching between tools / program stages / whatever. Disabling & enabling the pointer game objects is the most straightforward way to do this.

We should definitely keep the auto-populate option, but we should either leave it empty for now or make it opt-in. Before it can be useful we'll need to create an 'all-in-one' pointer that does both selection and teleportation, and which doesn't require any manual setup to work properly.

In the meantime I'll create an example scene where the pointer prefabs are just dragged into the scene and manually configured.

**Comment by Railboy** [comment]
I'm seeing now that the pointer base class no longer auto-registers itself with the InputManager, probably on the assumption that we would only be auto-populating them. That's another thing we'll have to bring back it we want to support custom setups.

**Comment by StephenHodgson** [comment]
> Coordinate the behavior of multiple interacting pointers. Eg, teleport pointers have an array called DisableWhileActive that can be populated with other pointers. This can prevent, say, the line pointer from being active while the teleportation pointer is active. But this relationship has to be set up explicitly. (This is why @cre8ivepark was seeing overlapping pointers.)

No, he was seeing the old teleport manager conflicting with the new, I believe.

**Comment by Railboy** [comment]
I'm referring to this, which I just saw myself:

>On thumbstick forward, the parabolic beam should be displayed without straight beam. (currently displayed together).

Hopefully this gives you a clearer picture of how multiple pointers will have to operate together. The goal now is to make it possible to drop an arbitrary number of pointer prefabs into the scene and have them play nicely with the FocusManager. Will you handle updating FocusManager to auto-register pointers in a way that plays well with your new input system? I added a bare-bones registration function as a starting point but the system has some expectations about input sources that I'm not sure how to fulfill.

**Comment by StephenHodgson** [comment]
> The goal now is to make it possible to drop an arbitrary number of pointer prefabs into the scene and have them play nicely with the FocusManager. Will you handle updating FocusManager to auto-register pointers in a way that plays well with your new input system?

Will do.  I thought I had something setup already when the pointer was activated to self register. I'll be sure to double check to make sure that didn't get changed/break.

**Comment by StephenHodgson** [comment]
@Railboy yeah, I changed the pointers to [self register as their sources are detected](https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/1792/files#diff-d6374fb44e8442d7c0eb90c6479f8363R771).

**Comment by StephenHodgson** [comment]
Let's switch back to opening PRs for changes to this branch, and not push directly to it.  I'd like to discuss the changes beforehand.

**Comment by Railboy** [comment]
Given how badly overdue these changes are and given how clear the mandate is I'd rather just power through this. We're two steps removed from the main branch (Input Refactor->Dev Branch->Main Branch) so devs are very well insulated. And you can feel free to make changes to my code without consulting me, I trust you to know the new input system better anyway.

I'd also like to divide the labor so that I'm sticking to pointer aesthetics as much as possible. If I ever hit a point where I'm not 100% sure how to proceed I'll stop, alert you and have you pick things up from there. Basically how we handled pointer registration. Work for you?

**Comment by StephenHodgson** [comment]
@Railboy Let's do this!

**Comment by Railboy** [comment]
I'm working on this as well. I'll be committing visual changes to the prefabs shortly.

>The Parabola arch logic needs some tuning to get the arch further from you when you point up.

I've added a flat animation curve and min/max velocity fields. By using the angle between up and the pointer's forward it interpolates between the min/max fields via the curve. Try messing with the values and see if you find something that feels right. I'll try on my end as well.

**Comment by cre8ivepark** [comment]
Could you please let me know how I can test it? MotionControllerTest scene gives me this error:
Assertion failed: This Pointer must have a Input Source Assigned
UnityEngine.Debug:Assert(Boolean, String)
MixedRealityToolkit.InputModule.Pointers.BaseControllerPointer:Start() (at Assets/MixedRealityToolkit/InputModule/Scripts/Pointers/BaseControllerPointer.cs:132)

I tried creating a new scene with MRTK configure menu but it does not give me motion controller visuals/input.

**Comment by StephenHodgson** [comment]
@cre8ivepark I think the MotionControllerTest scene is in a bad state, try any of the other scenes.

**Comment by Railboy** [comment]
@cre8ivepark Stephen is working on getting pointers to auto-register and to play nicely with the FocusManager. I believe this error message is related to that issue.

**Comment by cre8ivepark** [comment]
Tested with current status:

- There are three cursors in the scene. The Gaze cursor should disappear when a motion controller(s) is detected. (Gaze cursor should be displayed only when there are no motion controllers in the scene)

- On thumbstick forward, the parabolic beam should be displayed without straight beam. (currently displayed together).

- On thumbstick left/right/backward, the parabolic beam should not be displayed.
- Sometimes, only single controller visualizes the beam and cursor.
- Sometimes, a parabolic beam is displayed on trigger button event.
- Some NullReferenceExceptions
- Interesting bug: Only single pointe works in video recording mode. (I tried to capture video but multi-pointer is disabled on video capture)

mrtk_multipointer2
20180313_143326_mixedreality_moment 3

**Comment by StephenHodgson** [comment]
> Interesting bug: Only single pointer works in video recording mode. (I tried to capture video but multi-pointer is disabled on video capture)

Hmm, I wonder what could be causing this.  There's not really anything that could be interacting with it in such a way.

Thanks for the feedback, I turned your bullet points into a list to start checking off as we go.

**Comment by StephenHodgson** [comment]
Also, could you post any Null refs?

It'll be helpful to get the full stack trace as well.

**Comment by cre8ivepark** [comment]
@StephenHodgson I am getting these two messages:

NullReferenceException: Object reference not set to an instance of an object
MixedRealityToolkit.InputModule.Focus.FocusManager.**TryGetFocusDetails** (IPointer pointer, MixedRealityToolkit.InputModule.Focus.FocusDetails& focusDetails) (at Assets/MixedRealityToolkit/InputModule/Scripts/Focus/FocusManager.cs:325)
MixedRealityToolkit.InputModule.Cursors.BaseCursor.UpdateCursorTransform () (at Assets/MixedRealityToolkit/InputModule/Scripts/Cursors/BaseCursor.cs:350)
MixedRealityToolkit.InputModule.Cursors.BaseCursor.Update () (at Assets/MixedRealityToolkit/InputModule/Scripts/Cursors/BaseCursor.cs:278)

Assertion failed: **No Pointer has been assigned**!
UnityEngine.Debug:Assert(Boolean, String)
MixedRealityToolkit.InputModule.Cursors.BaseCursor:**UpdateCursorTransform()** (at Assets/MixedRealityToolkit/InputModule/Scripts/Cursors/BaseCursor.cs:347)
MixedRealityToolkit.InputModule.Cursors.BaseCursor:Update() (at Assets/MixedRealityToolkit/InputModule/Scripts/Cursors/BaseCursor.cs:278)

**Comment by davidkline-ms** [comment]
Quick note: There are many request for this in the current toolkit (master). We will need to revert then namespace change and ensure any changes required of the Input system are made to a master base.
- D

**Comment by StephenHodgson** [comment]
@davidkline-ms I really don't see this making its way into master very easily.  Not a good idea and way too much work.

**Comment by SimonDarksideJ** [comment]
This change is simply to breaking and has such a wide area of impact, it would not be prudent to apply in Master.  Would end up breaking existin projects using the toolkit.  Better to offer an upgrade path to the current Dev_Working if they need the features provided by the new input system, giving a cleaner implementation.

**Comment by davidkline-ms** [comment]
@StephenHodgson, @SimonDarksideJ,
Is there a minimal set of changes that support multi-pointer? What types of things do you expect to get broken in current projects?

Thanks!

**Comment by SimonDarksideJ** [comment]
@Railboy  may be able to provide more input on that from his initial work.

**Comment by StephenHodgson** [comment]
@davidkline-ms, just take a look at the changes in the PR.

**Comment by StephenHodgson** [comment]
Going to close this. Most of the work is already ported into vNEXT.
